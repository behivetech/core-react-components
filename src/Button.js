
// Vendor Libs
import React, {Component} from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import MdcButton from '@material/react-button';
import {pick} from 'lodash';

// Styles
import '@material/react-button/index.scss';

export default class Button extends Component {
    constructor(props) {
        super(props);
        this.handleClick = this.handleClick.bind(this);
    }

    handleClick(event) {
        event.preventDefault();

        this.props.onClick(event);
    }

    render() {
        const {children, className, type} = this.props;
        const props = {
            ...pick(this.props, [
                'children',
                'className',
                'dense',
                'disabled',
                'href',
                'icon',
                'outlined',
                'raised',
                'type',
                'unelevated',
            ]),
            children: children || type,
            className: classnames('button', className),
            onClick: this.handleClick,
        };
        
        return <MdcButton {...props} />;
    }
}

Button.propTypes = {
    /** Text to be displayed within root element. */
    children: PropTypes.string,
    /** Ability to add additional className to the component for addistional styling from parent */
    className: PropTypes.string,
    /** Enables dense variant and makes the button smaller */
    dense: PropTypes.bool,
    /** Disables button if true. */
    disabled: PropTypes.bool,
    /** Sets a hyperlink & uses anchor tag instead of a button. */
    href: PropTypes.string,
    /** Icon to render within root element. */
    icon: PropTypes.node,
    /** callback function to run when the button is clicked. */
    onClick: PropTypes.func,
    /** Enables outlined variant. */
    outlined: PropTypes.bool,
    /** Enables raised variant. */
    raised: PropTypes.bool,
    /** indicates the type of button it is. Defaults to button */
    type: PropTypes.oneOf(['button', 'reset', 'submit']),
    /** Enables unelevated variant. */
    unelevated: PropTypes.bool,
};

Button.defaultProps = {
    onClick: () => null,
    type: 'button',
};
