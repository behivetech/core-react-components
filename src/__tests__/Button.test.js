import Button from '../Button';

describe('Button', () => {
    afterEach(() => {
        jest.clearAllMocks();
    });

    it('should render with the correct props', () => {
        const wrapper = shallow(<Button />);
        const wrapperInstance = wrapper.instance();

        expect(wrapper).toMatchSnapshot();
        expect(wrapperInstance.props.onClick()).toBe(null);
        wrapper.setProps({className: 'mockClass', children: 'mock child'});
        expect(wrapper).toMatchSnapshot();
        wrapper.setProps({type: 'submit', children: undefined});
        expect(wrapper).toMatchSnapshot();
    });

    it('should handle clicks correctly', () => {
        const mockEvent = {
            preventDefault: jest.fn(),
        };
        const mockOnClick = jest.fn();
        const wrapperInstance = shallow(<Button onClick={mockOnClick} />).instance();

        wrapperInstance.handleClick(mockEvent);
        expect(mockEvent.preventDefault).toHaveBeenCalled();
        expect(mockOnClick).toHaveBeenCalledWith(mockEvent);
    });
});