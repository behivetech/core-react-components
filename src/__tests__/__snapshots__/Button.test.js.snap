// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render with the correct props 1`] = `
<RippledComponent
  className="button"
  disabled={false}
  initRipple={[Function]}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchStart={[Function]}
  style={Object {}}
  type="button"
  unbounded={false}
>
  button
</RippledComponent>
`;

exports[`Button should render with the correct props 2`] = `
<RippledComponent
  className="button mockClass"
  disabled={false}
  initRipple={[Function]}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchStart={[Function]}
  style={Object {}}
  type="button"
  unbounded={false}
>
  mock child
</RippledComponent>
`;

exports[`Button should render with the correct props 3`] = `
<RippledComponent
  className="button mockClass"
  disabled={false}
  initRipple={[Function]}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchStart={[Function]}
  style={Object {}}
  type="submit"
  unbounded={false}
>
  submit
</RippledComponent>
`;
